### PR TITLE
Update JS compiler content for v2.18

### PR DESCRIPTION
**Pull request by atsansone**
### Removed mentions of `dart2js` and `dartdevc`

* [FAQ](https://dart--staging.web.app/faq)
* [Null Safety FAQ](https://dart--staging.web.app/null-safety/faq)
* [Overview](https://dart--staging.web.app/overview)
* [`dart compile`](https://dart--staging.web.app/tools/dart-compile)
* [`dart`](https://dart--staging.web.app/tools/dart-tool)
* [tools](https://dart--staging.web.app/tools)
* [webdev](https://dart--staging.web.app/tools/webdev)
* [debugging](https://dart--staging.web.app/web/debugging)
* [deployment](https://dart--staging.web.app/web/deployment)

### Added production debugging section from `dart2js`
* [debugging](https://dart--staging.web.app/web/debugging)

### Removed all content
* [`dart2js`](https://dart--staging.web.app/tools/dart2js)
* [`dartdevc`](https://dart--staging.web.app/tools/dartdevc)

### CODE REVIEW

File: src/_guides/language/language-tour.md [hunk 1]
```diff
@@ -2470,9 +2470,9 @@ When exactly do assertions work?
 That depends on the tools and framework you're using:
 
 * Flutter enables assertions in [debug mode.][Flutter debug mode]
-* Development-only tools such as [dartdevc][]
+* Development-only tools such as [`webdev serve`][]
   typically enable assertions by default.
-* Some tools, such as [`dart run`][] and [dart2js][]
+* Some tools, such as [`dart run`][] and [`dart compile`][]
   support assertions through a command-line flag: `--enable-asserts`.
```
Comment (mit-mit) [inline src/_guides/language/language-tour.md:2476]: I don't think `dart compile` has the `--enable-asserts` flag?

File: src/tools/dart-compile.md [hunk 2]
```diff
@@ -136,11 +136,21 @@ plus a small [Dart runtime][] that handles
...
 ```terminal
 $ dart compile exe bin/myapp.dart -o /tmp/myapp
+```
+
+When successful, this command returns the following.
+```terminal
 Generated: /tmp/myapp
+```
+
+To run your compiled app, change to the `/tmp` directory and run.
```
Comment (mit-mit) [inline src/tools/dart-compile.md:152]: I think this would be simpler if it was just `/tmp/myapp` without having to change directory.

File: src/tools/dartdevc.md [hunk 3]
```diff
@@ -4,209 +4,15 @@ short-title: dartdevc
...
 {{site.alert.note}}
-  dartdevc is used for _development_ only.
-  Continue to use [dart2js](/tools/dart2js)
-  through [`dart compile js`](/tools/dart-compile#js) 
-  or the [`webdev`][] tool
-  to compile for deployment.
+  Dart 2.18 removes the `dartdevc` command-line tool from the Dart
+  package, but retains the dartdevc compiler.
+  
+  Use [`dart compile js`](/tools/dart-compile#js) to compile Dart code
```
Comment (mit-mit) [inline src/tools/dartdevc.md:15]: I don't think this is true. `compile js` will produce non-debuggable JS. I think maybe we need to point to webdev instead, @sigmundch ?

Comment (mit-mit) [reply src/tools/dartdevc.md:15]: It actually looks like the removed content below might all still work, and just need to be retained?

**Review by mit-mit — CHANGES_REQUESTED** [review]

File: src/_guides/language/language-tour.md (hunk 1 above)
Comment (atsansone) [reply src/_guides/language/language-tour.md:2476]: Maybe? See [dart2js code](https://github.com/dart-lang/sdk/blob/main/sdk/bin/dart2js#L40-L44).

File: src/tools/dartdevc.md (hunk 3 above)
Comment (atsansone) [reply src/tools/dartdevc.md:15]: This all falls under the `dartdevc` command. I can excise some parts and put them elsewhere, like `webdev`. I thought you were looking for removing anything other than `dart compile js`, so I tried to stay in that lane. Can you indicate which lines I should move? All of them, changing `dartdevc` to `webdev serve`?

File: src/tools/dart-compile.md (hunk 2 above)
Comment (atsansone) [reply src/tools/dart-compile.md:152]: Not a problem here.

File: src/_guides/language/language-tour.md (hunk 1 above)
Comment (sigmundch) [reply src/_guides/language/language-tour.md:2476]: We have it for the dart2js subcommand (`dart compile js`), but I don't know about all the other subcommands under `dart compile`. For example `dart compile exe` seems to have it, but jit-snapshot and kernel don't seem to have it.

File: src/_guides/language/language-tour.md [hunk 4]
```diff
@@ -4202,9 +4202,9 @@ Here are some cases when you might use deferred loading:
 * To load rarely used functionality, such as optional screens and dialogs.
 
 {{site.alert.warn}}
-  **Only dart2js supports deferred loading.**
-  Flutter, the Dart VM, and dartdevc don't support deferred loading.
-  For more information, see
+  **Only `dart compile js` supports deferred loading.**
```
Comment (sigmundch) [inline src/_guides/language/language-tour.md:4205]: FYI - I believe some support for deferred loading has been added to the VM and we are working towards adding it to the dartdevc workflow (webdev). That's tangential to this change, so I wouldn't change anything here, just wanted to give you a heads up :)

File: src/_guides/whats-new.md [hunk 5]
```diff
@@ -17,6 +17,17 @@ and follow the [Dart blog][].
...
+## August 30, 2022: 2.18 release
+
+This section lists notable changes made from May 12, 2022,
+through August 30, 2022.
+For details about the 2.18 release,
+see [Dart 2.18: ][], and the [change log][https://github.com/dart-lang/sdk/blob/main/CHANGELOG.md#2180].
+
+In addition to bug fixes and incremental improvements,
+we made the following changes to this site:
+
```
Comment (sigmundch) [inline src/_guides/whats-new.md:29]: is your intent to also add here some bullet points about these recent changes?

File: src/faq.md [hunk 6]
```diff
@@ -238,16 +238,16 @@ long enough for runtime optimizations to be applied.
...
 
   * Chrome
   * Edge
   * Firefox
   * Safari
 
-The _development_ compiler ([dartdevc][]) supports only Chrome.
-You _might_ be able to use other modern browsers
+The _development_ web compiler supports Chrome only.
+You _might_ be able to use other modern browsers based on Chromium
```
Comment (sigmundch) [inline src/faq.md:250]: I'd recommend reverting this line change - only Edge is based on chromium, FF and Safari are not.

File: src/faq.md [hunk 7]
```diff
@@ -238,16 +238,16 @@ long enough for runtime optimizations to be applied.
...
 (Chromium Edge, Firefox, and Safari).
```
Comment (sigmundch) [inline src/faq.md:251]: Consider removing there "Chromium", since now the default is that Edge is a chromium based browser?

File: src/tools/dartdevc.md (hunk 3 above)
Comment (sigmundch) [reply src/tools/dartdevc.md:15]: My recommendation here is that we are not downplaying one command over others. Both commands are serving different purposes and each is valid for the journey the developer is at the moment (I provided additional details in the planning doc if that helps).

That means that we'd keep both references to web development and web production compilers, but change the sample commands. If a command used `dart2js`, we turn it into `dart compile js`, if a command used `dartdevc`, we turn it into `webdev serve`.

In the past, calling `dartdevc` directly was not recommended anyways, and not something I expect we had said to do in the documentation (we used the term to refer to the development compiler, but I'd be surprised if we had examples to show people how to call it directly). So I expect for the most part we used `webdev serve` to describe those use cases, and we can simply preserve good chunks of the old documentation.

File: src/null-safety/faq.md [hunk 8]
```diff
@@ -319,72 +319,75 @@ and situations where a null value is really expected. So the tool tells you what
 it knows ("it looks like this condition will always be false!") and lets you
 decide what to do.
 
-## What should I know about dart2js and null safety?
+## What should I know about compiling and null safety?
 
-For a long time, the dart2js compiler has had optimizations specifically
-targeting null values and null checks. Because of that, null safety is not
-expected to affect much the output of the compiler.
+The compiler optimizes null values and null checks. So null safety
```
Comment (sigmundch) [inline src/null-safety/faq.md:324]: I think there is some missing pieces in this paragraph that it's worth adding.  I realize now that when I first wrote this I didn't word this properly to really convey what I was trying to say, so while your edits make sense looking at this with fresh eyes, I feel that some of the nuance that I was trying to convey here gets lost in the process.

To provide additional context. Developers know that null safety bring many benefits. One such benefit is that it will reduce code-size and improve performance of apps compiled natively (e.g. flutter and AOT targets). Many developers believe that the same benefits will come to JavaScript apps, but in reality the benefits will be much smaller. This Q/A was added because we wanted to set expectations appropriately.

My goal with the first sentence "For a long time ..." was to provide the reason behind this. That is, while our compiler learns from null-safety and it makes the compiler job easier, it's just that in the past there were many optimizations added to the production web compiler to specifically recover the cost of null checks, and as a result the benefits of null safety are not as easily observable when looking at the output of the compiler.

There is probably a better way to convey all of this :). I know this is tangential to your edits, so I'd be fine keeping the original wording for now.

File: src/null-safety/faq.md [hunk 9]
```diff
@@ -319,72 +319,75 @@ and situations where a null value is really expected. So the tool tells you what
...
-  unnecessary null checks when it knows the value is not null.
-
-* By default dart2js would generate parameter subtype checks (runtime checks
-  used to ensure covariant virtual calls are given appropriate arguments).
-  These are elided with the `--omit-implicit-checks` flag just as before.
-  Recall that this flag can make programs have unexpected behavior if types
-  are invalid, so we continue to recommend that the code has strong test
-  coverage to avoid any surprises. In particular, dart2js optimizes code based
+* The compiler generates `!` null assertions, but you might not notice
+  them. Pre null-safety, the compiler generated null checks that still
```
Comment (sigmundch) [inline src/null-safety/faq.md:330]: nit: reading this with fresh eyes - this is not very easy to follow. Maybe the subject here is to implicit (my fault when I first added this text, I think). Suggestion:

* The production web compile generates `!` null assertions, but you might not notice them when comparing the output of the compiler before and after adding null assertions. That's because the compiler already emitted null checks in programs that weren't null-safe.

File: src/null-safety/faq.md [hunk 10]
```diff
@@ -319,72 +319,75 @@ and situations where a null value is really expected. So the tool tells you what
...
+  remain in existing programs).
+
+* The compiler generates these null assertions regardless of the
+  soundness of null safety or optimization level. In fact, the compiler
+  doesn't remove `!` when using `-O3` or `--omit-implicit-checks`.
+
+* The compiler might remove unnecessary null checks. Previous compiler
```
Comment (sigmundch) [inline src/null-safety/faq.md:337]: I have a similar concern here as with the the intro paragraph: I think the context is getting lost. Seeing your edit reveal to me that the original wording was vague, so it highlights that we need to be more explicit about that context.

A suggestion here may be to clarify that when we used to say "optimizations done in the past" we meant "optimizations done in the compiler prior to null-safety".

File: src/null-safety/faq.md [hunk 11]
```diff
@@ -319,72 +319,75 @@ and situations where a null value is really expected. So the tool tells you what
...
-  types before, and is true with inconsistent nullabilities now with sound
-  null-safety.
+  the code provides arguments of an invalid type, those optimizations
+  would be wrong and the program could misbehave. This was true for
+  inconsistent types before, and is true with inconsistent 
+  nullabilities now with sound null-safety.
 
-* You may notice that DDC and the Dart VM have special error
-  messages for null checks, but to keep applications small dart2js does not.
+* You may notice that the development mode of the compiler and the Dart
```
Comment (sigmundch) [inline src/null-safety/faq.md:354]: nit: replace "the development mode of the compiler" with "the web development compiler".

File: src/overview.md [hunk 12]
```diff
@@ -187,9 +187,8 @@ Dart's compiler technology lets you run code in different ways:
   an ahead-of-time (AOT) compiler for producing machine code.
 
 * **Web platform**: For apps targeting the web,
-  Dart includes both a development time compiler (dartdevc) and
-  a production time compiler (dart2js).
-  Both compilers translate Dart into JavaScript.
+  Dart can compile for debugging or production.
```
Comment (sigmundch) [inline src/overview.md:190]: nit: debugging/development ? (to keep things consistent?)

File: src/overview.md [hunk 13]
```diff
@@ -187,9 +187,8 @@ Dart's compiler technology lets you run code in different ways:
...
+  Its web compiler translates Dart into JavaScript.
```
Comment (sigmundch) [inline src/overview.md:191]: Technically there are 2 ocmpilers though... another thought here is to keep the original wording, but remove the words in parenthesis from the original sentence?

File: src/tools/dart-compile.md [hunk 14]
```diff
@@ -259,24 +254,45 @@ they can have much slower startup than architecture-specific AOT output formats.
...
 
-Here's an example of compiling a Dart application to JavaScript
-with many optimizations enabled:
+{{site.alert.note}}
+  Use the [`webdev` tool][webdev] rather than running the 
+  Dart-to-JavaScript compiler.
+  The [`webdev build`][] command, by default,
+  also produces minified, deployable JavaScript.
+  The [`webdev serve`][] command, by default,
+  produces JavaScript modules to compile web apps
```
Comment (sigmundch) [inline src/tools/dart-compile.md:263]: nit: delete "to compile web apps" ?

File: src/_includes/tools/dart-compile-js-options.md [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+  run `dart compile js -hv`.
+
+  * `-O0`: Disables many optimizations.
+  * `-O1`: Enables default optimizations.
+  * `-O2`: Enables `-O1` optimizations, plus additional ones
+    (such as minification) that respect the language semantics and
+    are safe for all programs.
+    {{site.alert.note}}
+      With `-O2`, string representations of types are no longer the same as
+      those in the Dart VM and when compiled with [dartdevc][].
```
Comment (sigmundch) [inline src/_includes/tools/dart-compile-js-options.md:34]: note: one more reference to dartdevc to remove here.

File: src/tools/dartdevc.md (hunk 3 above)
Comment (sigmundch) [reply src/tools/dartdevc.md:15]: Correct - in this case we'd point them at `webdev serve`

There is still a page for webdev below, so rather than retaining the content of the dartdevc.md file, I'd recommend selecting the relevant pieces to preserve that make sense to show in the context of webdev. From looking at the contents on this file, thought, I don't think there is much I'd keep (all content about module format, etc, are not really that relevant to our developers).

File: src/web/deployment.md [hunk 16]
```diff
@@ -3,67 +3,31 @@ title: Web deployment
...
-  If some code paths aren't tested,
-  your app might run with dartdevc but
-  misbehave when compiled using dart2js.
-{{site.alert.end}}
-
-You can specify dart2js options in a build config file,
-as described in the [build_web_compilers page.][build_web_compilers]
+[Use the `webdev build` command][build] to create a deployable version
+of your app. This command converts your code to JavaScript and saves
+the result as `build/web/main.dart.js`.
```
Comment (sigmundch) [inline src/web/deployment.md:25]: I agree with most of the deletions here, but it may be worth including also a link to the dart2js options (I wouldn't document them individually here, but refer to the fact that under the hood `webdev build` invokes `dart compile js`, so all options available there can also be used to configure a `webdev build`.

File: src/tools/dart-compile.md (hunk 2 above)
Comment (sfshaza2) [reply src/tools/dart-compile.md:152]: Agree with mit.

File: src/_guides/whats-new.md (hunk 5 above)
Comment (parlough) [reply src/_guides/whats-new.md:29]: I guess this was accidentally included as that is being handled in https://github.com/dart-lang/site-www/pull/4195

File: src/overview.md (hunk 13 above)
Comment (atsansone) [reply src/overview.md:191]: I made a different update. Let me know what you think.

File: src/tools/dartdevc.md [hunk 17]
```diff
@@ -4,209 +4,15 @@ short-title: dartdevc
...
-  dartdevc is used for _development_ only.
-  Continue to use [dart2js](/tools/dart2js)
-  through [`dart compile js`](/tools/dart-compile#js) 
-  or the [`webdev`][] tool
-  to compile for deployment.
+  Dart 2.18 removes the `dartdevc` command-line tool from the Dart
+  package, but retains the dartdevc compiler.
+  
+  Use [`webdev serve`](/tools/webdev#serve) to compile Dart code
+  to modular JavaScript. You can debug these modules more easily.
```
Comment (johnpryan) [inline src/tools/dartdevc.md:16]:
```suggestion
  Use [`webdev serve`](/tools/webdev#serve) to compile Dart code
  to modular JavaScript. This makes compilation faster and debugging easier.
```

File: src/faq.md [hunk 18]
```diff
@@ -238,17 +238,17 @@ long enough for runtime optimizations to be applied.
...
 
   * Chrome
   * Edge
   * Firefox
   * Safari
 
-The _development_ compiler ([dartdevc][]) supports only Chrome.
-You _might_ be able to use other modern browsers
-(Chromium Edge, Firefox, and Safari).
+The _development_ web compiler supports Chrome only. You _might_ be
```
Comment (johnpryan) [inline src/faq.md:249]: Should we link to webdev here?

File: src/faq.md [hunk 19]
```diff
@@ -373,34 +370,38 @@ For example, the `dart:io` library
...
+right compiler for your use case. 
 
-The two compilers have different use cases. You don't usually have to worry
-about which compiler you're using, because the [webdev][] tool
-chooses the right compiler for your use case. When you're developing your app,
-webdev chooses [dartdevc][], which supports incremental compilation so
-you can quickly see the results of your edits.
-When you're building your app for deployment, webdev chooses [dart2js][],
-which uses techniques such as tree shaking to produce optimized code.
+* When you're developing your app, `webdev` chooses a development web
```
Comment (johnpryan) [inline src/faq.md:380]:
```suggestion
* When you're developing your app, `webdev` uses Dart's development
```

File: src/faq.md [hunk 20]
```diff
@@ -435,19 +436,15 @@ staying within smi range is still good practice.
...
-Float32List becomes a Float32Array.
-The one exception today is that dart2js does not support 64-bit integers
-and thus does not support Int64List or Uint64List.
-Dart code compiled via dart2js results in a runtime exception
-if either of those lists is used.
+JavaScript offers typed arrays compatible with Dart’s typed lists.
+The mapping is trivial: `Float32List` becomes a `Float32Array`.
+One exception exists: the production web compiler does not support
+64-bit integers: `Int64List` or `Uint64List`. Compiling Dart code with
+either of those lists results in a runtime exception.
```
Comment (johnpryan) [inline src/faq.md:443]: Just FYI, [semantic line breaks](https://sembr.org/) like these are a convention the Dart and Flutter sites often use

File: src/null-safety/faq.md [hunk 21]
```diff
@@ -172,13 +172,16 @@ The assert will be unnecessary when everything is fully migrated, but for now it
...
-But, the check *is* still needed if the program is a mixed-version one.  Until
-everything is fully migrated and the code switches to running with sound null
-safety, it will be possible for `arg` to be null.
+```dart
+if (arg == null) throw ArgumentError(...)`
+```
+
+You must include this check if the program is a mixed-version one.
+Until everything is fully migrated and the code switches to running
+with sound null safety, `arg` might be set to `null`.
```
Comment (johnpryan) [inline src/null-safety/faq.md:184]: This is unrelated to js compilation, not sure if this was committed by accident

File: src/tools/index.md [hunk 22]
```diff
@@ -103,23 +102,12 @@ The Dart SDK includes the following general-purpose `dart` tool:
 : A suite of debugging and performance tools.
 
 
-## Tools for developing web apps {#web}
+## Tool for developing web apps {#web}
 
 The following tools support developing web apps:
```
Comment (johnpryan) [inline src/tools/index.md:107]: This section might need a bigger update, but for consistency: 

```suggestion
The following tool support developing web apps:
```

File: src/tools/webdev.md [hunk 23]
```diff
@@ -4,68 +4,76 @@ description: Command-line tools for Dart web development.
...
+`build_runner` to test your app.
 
 ## Setup
 
-The easiest way to get `webdev` is to globally install it,
-so that it can be [in your PATH.][PATH]
-Before you can use `webdev`,
-your web app must depend on the
-build_runner and build_web_compilers packages.
+Install `webdev` for all users. This adds the tool [to your PATH][PATH].
```
Comment (johnpryan) [inline src/tools/webdev.md:12]:
```suggestion
Follow these instructions to get started using `webdev`.
```

File: src/tools/webdev.md [hunk 24]
```diff
@@ -4,68 +4,76 @@ description: Command-line tools for Dart web development.
...
-to your app's `pubspec.yaml` file:
+To depend on these packages, add the following [dev dependencies][] to
+your app's `pubspec.yaml` file:
 
 <!--?code-excerpt "quickstart/pubspec.yaml (build dependencies)"?-->
-```
+```yaml
   dev_dependencies:
-    # ···
+    ···
```
Comment (johnpryan) [inline src/tools/webdev.md:48]: The hash is a YAML comment, I'm not sure if we validate the yaml snippets or not, but if we ever do, we should avoid causing trouble:

```suggestion
    # ···
```

File: src/tools/dartdevc.md (hunk 3 above)
Comment (sigmundch) [reply src/tools/dartdevc.md]: Is it a requirement that we keep old pages around? or is it OK to delete them altogether?

After looking at this more closely, I'm no longer sure I see the value in keeping around both the dartdevc.md and the dart2js.md pages. So I'd be inclined in deleting them both. If we have to keep them, then I would suggest to trim them down further. For example, by deleting anything that is describing the compilers beyond production/development (e.g. I'd suggest deletion the whole explanation of _modules_ and _AMD_).

One suggestion:
* _"Version note: Since Dart 2.18 the `dart2js` web production compiler is only available from `dart compile js`(link)"_, and
* _"Version note: Since Dart 2.18 the `dartdevc` web development compiler is only available from `webdev serve`(link)."_ 

Another suggestion (I slightly prefer this one):
* _"dart2js is Dart's web production compiler. Version note: Since Dart 2.18 the `dart2js` compiler can only be invoked from `dart compile js`. It is also used internally by deployment tools like `webdev build`"_, and
* _"dartdevc is Dart's web development compiler. Version note: Since Dart 2.18 the `dartdevc` compiler cannot be invoked directly, but it is used internally by development tools like  `webdev serve`."_

File: src/tools/webdev.md (hunk 24 above)
Comment (parlough) [reply src/tools/webdev.md:48]: Yeah I suppose we added the hash in-case the user wanted to copy this snippet directly. 

Perhaps we should switch these instructions to `dart pub add` instead anyway though as those mentioned versions get out of date.

File: src/overview.md (hunk 13 above)
Comment (sigmundch) [reply src/overview.md]: How about "Its web compilers translate Dart into JavaScript" (plural compilers vs singluar compiler?)

Alternatively, combining both sentences:
Web platform: For apps targeting the web, Dart can compile to JavaScript, both for development and production purposes.

File: src/tools/index.md [hunk 25]
```diff
@@ -103,23 +102,12 @@ The Dart SDK includes the following general-purpose `dart` tool:
...
-[dart2js](/tools/dart2js)
-: The original Dart-to-JavaScript compiler, with tree shaking.
-  IDEs, `dart compile js`, and the `webdev` CLI 
-  use dart2js when building web apps for deployment.
-
-[dartdevc](/tools/dartdevc)
-: The Dart dev compiler, a modular Dart-to-JavaScript compiler.
-  IDEs and the `webdev` CLI use dartdevc when running a development server.
-
+: A CLI to build and serve Dart web apps.
```
Comment (sigmundch) [inline src/tools/index.md:110]: Should we consider adding back a bullet point equivalent to `dart2js` (maybe it would be `dart compile js` instead?).

It seems these were the primary links we had to steer readers towards the more detailed articles about how to develop and deploy apps to JavaScript, so without those, this index page feels incomplete.

Also worth nothing something for a follow up (clearly out of scope for this change): it's strange that have a couple workflow articles like web/deployment and web/debugging, but no article on web/development. It would be be consistent and have an article per user journey. My guess the reason is that it overlaps entirely with the content of the webdev.md article, so it would require a bit of editing to figure out the right organization of that content.

**Review by sigmundch — COMMENTED** [review]
Overall the changes look great, thanks @atsansone!

I defer to the proper editors here to give the final stamp of approval 😊

**Comment by mit-mit** [comment]
@atsansone is this ready for a last review? If so, could you please stage it somewhere? I'm having a bit of trouble visualizing how the pages end up linking together from just the source.

**Comment by atsansone** [comment]
> 

@mit-mit : The links to the staged pages are in the description at the top of this PR.

File: src/_includes/known-issues/compile-ki.md [hunk 26]
```diff
@@ -0,0 +1,20 @@
+No cross-compilation support ([issue 28617][])
```
[the remark this answers is not in the document]
Comment (atsansone) [reply src/_includes/known-issues/compile-ki.md:1]: I yanked out this block from the `dart compile` file to cover `exe` and `aot-snapshot` more cleanly. It is unrelated. Apologies for the extra change.

File: src/_includes/tools/dart-compile-js-options.md [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+
+Some other handy options include:
+
+`--packages=<path>`
+: Specifies the path to the package resolution configuration file.
+  For more information, see
+  [Dart Package Configuration File.][]
+
+`-D<flag>=<value>`
+: Defines an environment variable.
```
Comment (kevmoo) [inline src/_includes/tools/dart-compile-js-options.md:66]: We have to have a doc on this that can be linked to.

I'm annoyed this page https://api.dart.dev/stable/2.18.0/dart-core/int/int.fromEnvironment.html – and others don't link to something like that!

File: src/faq.md (hunk 20 above)
Comment (atsansone) [reply src/faq.md:443]: Just learned what this is. Great point. I'll fix some here and others later.

File: src/null-safety/faq.md (hunk 21 above)
Comment (atsansone) [reply src/null-safety/faq.md:184]: This was to clean up the language and format around this check. It's the same content as what was there originally.

File: src/tools/index.md (hunk 25 above)
Comment (atsansone) [inline src/tools/index.md:110]:
```suggestion
: A CLI to [build](/tools/webdev#build) and [serve](/tools/webdev#serve) Dart JavaScript apps.
```

**Review by atsansone — COMMENTED** [review]
Made corrections per @johnpryan .

File: src/_guides/language/language-tour.md [hunk 28]
```diff
@@ -2470,9 +2470,9 @@ When exactly do assertions work?
 That depends on the tools and framework you're using:
 
 * Flutter enables assertions in [debug mode.][Flutter debug mode]
-* Development-only tools such as [dartdevc][]
+* Development-only tools such as [`webdev serve`][]
```
Comment (mit-mit) [inline src/_guides/language/language-tour.md:2473]: This like and the compile js one below don't render correctly. I think you are missing definitions of them.

Screen Shot 2022-08-30 at 10 49 31

**Review by mit-mit — APPROVED** [review]
Just one more comment (left above)

**Review by kevmoo — APPROVED** [review]

File: src/_includes/tools/dart-compile-js-options.md [hunk 29]
```diff
@@ -0,0 +1,113 @@
+#### Options {#prod-compile-options}
```
Comment (johnpryan) [inline src/_includes/tools/dart-compile-js-options.md:1]: includes are typically not used for sections like this. This can make things harder to maintain (the heading type might change depending on the context where it's included.) IMO it's better to just inline things and use includes for banners and other smaller things.

**Review by johnpryan — APPROVED** [review]
